DTO validation

Nitoriubar opened the pull request:
## 📝 세부사항
- 전체적으로 DTO들의 validation을 하여 Request의 값들에 대한 유효성 검사를 함
- 정규표현식을 사용하여 Custom Validation으로 직접 validation을 구현함
- 팔로우 CRUD 기능 구현
- `FollowUserRequest`의 `Id`값을 Validation함
- 로그인한 사용자들만 팔로워/팔로잉을 볼 수 있음
- 팔로워/팔로잉 리스트를 `FollowRepository`의 `findAllByFollowing`, `findAllByFollower`를 통해 가져옴
- 현재 사용자가 다른 사람의 팔로워/팔로잉 리스트를 확인할 경우 `isFollow`메서드로 그 중 현재 사용자를 팔로워한 사용자가 있는지 확인함

## 📗 참고사항
- [Custom Validation](https://www.baeldung.com/java-bean-validation-constraint-composition)
- [Validation의 내부동작원리](https://www.adrianbartnik.de/blog/annotation-processing/)
- [정규표현식 테스트 사이트](https://regexr.com/)

## 📎 관련 이슈
- #38
- #40

Nitoriubar committed:
Dto에 Validation 추가 완료

Nitoriubar committed:
Dto에 Validation 추가 완료

Nitoriubar committed:
reply의 DTO들에 Validation추가 완료

Nitoriubar committed:
Post의 DTO들에 Validation추가 완료

Nitoriubar committed:
trip의 DTO들에 Validation추가 완료

Nitoriubar committed:
user의 DTO들에 Validation추가 완료

Nitoriubar committed:
searchDto들 validation 완료

Nitoriubar committed:
replyDto들 Validation완료

Nitoriubar committed:
postDto들 Validation완료

Nitoriubar committed:
postGroupDto들 Validation완료

Nitoriubar committed:
userDto Validation완료

Nitoriubar committed:
feat:MyprofileController에 followUser 및 FollowUserRequest 추가함

Nitoriubar committed:
MyProfileService 및 User에 followUser 추가하여 follow구현함

Nitoriubar committed:
feat: User의 followUser parent설정을 추가함

Nitoriubar committed:
User에서 Follow패키지 생성하여 Follow기능을 분리함

sangrimlee commented on src/main/java/com/triprint/backend/core/valid/LocationValidator.java:
```diff
@@ -0,0 +1,29 @@
+
+	@Override
+	public boolean isValid(String location, ConstraintValidatorContext context) {
+		String regex = "^[-+]?[0-9]+(?:.[0-9]+)?$";
```
제가 이 부분을 작성할 때는 변수 이름을 생각하지 않고, 그냥 대충 변수 이름 작성했었지만, 정규표현식 같은 경우에 해당 정규표현식을 한 번에 이해하기 힘들기 때문에, 어떠한 정규표현식인지 명확하게 변수명을 지어주시면 좋아요!

Nitoriubar replied:
넵! 
피드백을 바탕으로 수정하겠습니다! 감사합니다. :)

sangrimlee commented on src/main/java/com/triprint/backend/domain/search/dto/CurrentLocationRequest.java:
```diff
@@ -1,23 +1,27 @@
-	@Range(min = -180, max = 180)
+	// 얘는 경도 180
+	@ValidLocation(message = "올바른 x값을 입력해주세요", locationType = LocationType.LONGITUDE)
 	private String x;
 
-	@NonNull
-	@Range(min = -90, max = 90)
+	// 얘는 위도 90
```
이러한 주석을 [javadoc](https://docs.oracle.com/javase/8/docs/technotes/tools/windows/javadoc.html)을 사용해서 `@ValidLocation`에 주석을 남겨주시면 좋아요.

그러면 ValidLocation을 사용할 때에, 위도, 경도 즉 `LocationType`enum에 따른 범위를 명시해주면, 좋습니다.

Nitoriubar replied:
넵, 수정하겠습니다!

sangrimlee reviewed (approved):
고생하셨습니다 👍

Nitoriubar committed:
Follow Entity 생성 및 FollowRepository, FollowService를 생성해 팔로우 기능 구현

Nitoriubar committed:
FollowController에 unfollow구현완료

Nitoriubar committed:
Follow가 존재하지않을 경우 ErrorMessage 생성함

Nitoriubar committed:
FollowEntity에 unfollow메서드를 생성하여 각 유저들의 팔로워|팔로잉 리스트에서 삭제한 후 FollowService에서 해당 Follow를 삭제하도록 UnFollow기능 구현

Nitoriubar committed:
feat: 팔로워 리스트 조회 기능 구현
- FollowerController에 Getmapping으로 리스트 조회 기능 추가함
- GetFollowersResponse Dto 추가함

Nitoriubar committed:
fix: 언팔로워 기능 수정
- Follow Entity에서 unfollow메서드 삭제하고 User Entity에서 casecade로 변경함
- FollowerService에서 unfollow를 수정함

Nitoriubar committed:
feat: 팔로워 리스트 가져오기 기능 구현 완료

Nitoriubar committed:
feat: 팔로잉 리스트 기능 추가
- 현재 로그인한 유저가 팔로우 한 사람들의 리스트를 Pagination으로 가져옴

Nitoriubar committed:
fix: 팔로우의 전반적인 로직 수정
- Follow엔티티의 follower: 팔로우를 하는 사람/ following: 팔로우를 받는 사람
- Follow엔티티에 맞도록 FollowService의 Follow, Unfollow수정함
- FollowService의 getFollowers: '나'를 팔로우한 사람들/getFollowings: '나'의 팔로우를 받는 사람들 에 맞도록 로직 수정함

Nitoriubar committed:
feat: 다른사람의 팔로워 / 팔로잉 리스트 가져오는 기능 구현
- 다른 사람의 팔로워 / 팔로잉 리스트 가져오기 기능 구현 완료
- 현재 사용자가 팔로우하고 있는 사용자가 다른 사람의 팔로워 / 팔로잉 리스트에 존재한다면 isFollow를 true로 하여 반환해주는 기능 구현중

Nitoriubar committed:
feat: 다른 사람의 팔로워 / 팔로잉 목록 기능 추가
- 현재 로그인한 사용자가 다른 사용자의 팔로잉/팔로워 목록에 있는 사람을 팔로잉 or 팔로워 하고 있다면 표시함
- Controller에서 follow/following/3 일 경우 3번을 팔로잉 하는 유저를 가져옴
- Controller에서 follow/follower/3 일 경우 3번을 팔로워 하는 유저를 가져옴

Nitoriubar committed:
refactor: User정보 가져오는 DTO 수정 및 GetFollowResponse구조 변경함
- User정보 가져오는 Dto를 사용하는 곳이 많기에 이름을 변경하여 어디에든 적용할 수 있도록 수정함
- GetFollowResponse DTO에서 Follower안에 isFollow가 있는게 더 나아보여서 해당사항을 적용하여 수정함

Nitoriubar committed:
style: 변수명 수정
- 위도와 경도에 관한 정규식이므로 변수명을 명확하게 변경함

Nitoriubar committed:
refactor: 위/경도의 CustomAnnotation에 주석을 추가함
- ValidLocation사용시 위/경도에 따른 범위를 주석으로 명시해줌

Nitoriubar committed:
style: 주석위치 변경
- ValidLocation에 해당 주석이 나와야하므로 위치를 변경함
- LocationType이 필요하지 않기에 삭제함
